File: qpid-proton-cpp/servers/receive.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include <proton/connection.hpp>
#include <proton/container.hpp>
#include <proton/delivery.hpp>
#include <proton/listener.hpp>
#include <proton/message.hpp>
#include <proton/messaging_handler.hpp>
#include <proton/receiver.hpp>

#include <iostream>
#include <string>

struct receive_handler : public proton::messaging_handler {
    std::string listen_url_ {};
    std::string address_ {};
    int desired_ {0};
    int received_ {0};

    proton::listener listener_ {};

    void on_container_start(proton::container& cont) override {
        listener_ = cont.listen(listen_url_);

        std::cout << "RECEIVE: Listening at " << listen_url_ << "\n";
    }

    void on_message(proton::delivery& dlv, proton::message& msg) override {
        std::cout << "RECEIVE: Received message '" << msg.body() << "'\n";

        received_++;

        if (received_ == desired_) {
            listener_.stop();
            dlv.receiver().close();
            dlv.connection().close();
        }
    }
};

int main(int argc, char** argv) {
    if (argc != 3 && argc != 4) {
        std::cerr << "Usage: receive <connection-url> <address> [<message-count>]\n";
        return 1;
    }

    receive_handler handler {};
    handler.listen_url_ = argv[1];
    handler.address_ = argv[2];

    if (argc == 4) {
        handler.desired_ = std::stoi(argv[3]);
    }

    proton::container cont {handler};

    try {
        cont.run();
    } catch (const std::exception& e) {
        std::cerr << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: qpid-proton-python/servers/receive.py
#!/usr/bin/python
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from __future__ import print_function

import sys

from proton.handlers import MessagingHandler
from proton.reactor import Container, ReceiverOption

class ReceiveHandler(MessagingHandler):
    def __init__(self, listen_url, address, desired):
        super(ReceiveHandler, self).__init__()

        self.listen_url = listen_url
        self.address = address

        self.desired = desired
        self.received = 0

        self.acceptor = None

    def on_start(self, event):
        self.acceptor = event.container.listen(self.listen_url)

        print("RECEIVE: Listening at {0}".format(self.listen_url))

    def on_link_opening(self, event):
        print("RECEIVE: Opening receiver for target address '{0}'".format
              (event.receiver.remote_target.address))

        # Set the target address using the value from the remote peer
        address = event.receiver.remote_target.address
        event.receiver.target.address = address

    def on_message(self, event):
        print("RECEIVE: Received message '{0}'".format(event.message.body))

        self.received += 1

        if self.received == self.desired:
            self.acceptor.close()
            event.connection.close()

def main():
    try:
        listen_url, address = sys.argv[1:3]
    except ValueError:
        sys.exit("Usage: receive.py <connection-url> <address> [<message-count>]")

    try:
        desired = int(sys.argv[3])
    except (IndexError, ValueError):
        desired = 0

    handler = ReceiveHandler(listen_url, address, desired)
    container = Container(handler)
    container.run()

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass
